### PR TITLE
fix: from csv behaviour

### PR DESCRIPTION
**Pull request by anna-charlotte**
Goals:

- Solving this issue: https://github.com/jina-ai/docarray/issues/626
- [ ] check and update documentation, if required. See [guide](https://github.com/jina-ai/jina/blob/master/CONTRIBUTING.md#documentation-guidelines)

### CODE REVIEW

**Comment by JoanFM** [comment]
I believe this can be solved inside `from_csv` itself checking if `cls` is an Instance or a class.

File: docarray/array/mixins/io/from_gen.py [hunk 1]
```diff
@@ -129,6 +134,17 @@ def from_csv(cls: Type['T'], *args, **kwargs) -> 'T':
         """
         return cls._from_generator('from_csv', *args, **kwargs)
 
+    def __from_csv(cls: Type['T'], *args, **kwargs) -> 'T':
+        """
+        # noqa: DAR101
+        # noqa: DAR102
+        # noqa: DAR201
+        """
+        warnings.warn(
```
Comment (JoanFM) [inline docarray/array/mixins/io/from_gen.py:143]: I am not sure if we should raise a `NotExistingMethod` exception or something like that @samsja , something that is the same as what `pandas` do.

Comment (samsja) [reply docarray/array/mixins/io/from_gen.py:143]: Yes I agree as well that we should raise an error rather than a warning

**Comment by JohannesMessner** [comment]
> Do not add an `init` to the `Mixin`. DO that in the `init` of DocArray Base class

Are we sure about this? It would assign some methods that don't exist there and break separation of concerns.

**Comment by JoanFM** [comment]
> > Do not add an `init` to the `Mixin`. DO that in the `init` of DocArray Base class
> 
> Are we sure about this? It would assign some methods that don't exist there and break separation of concerns. It seems to be accepted practice to let mixins have inits and pass **kwargs between them: https://stackoverflow.com/a/50465583

multiple inheritance is not so easily supported in Python I think. What is ur super when u call super?

**Comment by JohannesMessner** [comment]
> > > Do not add an `init` to the `Mixin`. DO that in the `init` of DocArray Base class
> > 
> > 
> > Are we sure about this? It would assign some methods that don't exist there and break separation of concerns. It seems to be accepted practice to let mixins have inits and pass **kwargs between them: https://stackoverflow.com/a/50465583
> 
> multiple inheritance is not so easily supported in Python I think. What is ur super when u call super?

My understanding is that it picks a certain evaluation order along which it tries to find an __init__ to call, and then it call the first it finds.
But if that __init__ in turn calls `super().__init__(**kwargs)` then it will keep 'walking' along that evaluation order. So as long as every init in a mixin makes that call it should all work as intended. So the mixins just need to be set up to not break this.

**Comment by samsja** [comment]
> > > > Do not add an `init` to the `Mixin`. DO that in the `init` of DocArray Base class
> > > 
> > > 
> > > Are we sure about this? It would assign some methods that don't exist there and break separation of concerns. It seems to be accepted practice to let mixins have inits and pass **kwargs between them: https://stackoverflow.com/a/50465583
> > 
> > 
> > multiple inheritance is not so easily supported in Python I think. What is ur super when u call super?
> 
> My understanding is that `super()` picks a certain evaluation order along which it tries to find an `__init__` to call, and then it call the first it finds. But if that `__init__` in turn calls `super().__init__(**kwargs)` then it will keep 'walking' along that evaluation order. So as long as every init in a mixin makes that call it should all work as intended. The mixins just need to be set up to not break this chain.

I agree with Johannes it would be better to add it in the Mixin `__init__` but if it is not possible because of inheritance problem we can put in in the DA __init__

**Comment by JoanFM** [comment]
> > > > > Do not add an `init` to the `Mixin`. DO that in the `init` of DocArray Base class
> > > > 
> > > > 
> > > > Are we sure about this? It would assign some methods that don't exist there and break separation of concerns. It seems to be accepted practice to let mixins have inits and pass **kwargs between them: https://stackoverflow.com/a/50465583
> > > 
> > > 
> > > multiple inheritance is not so easily supported in Python I think. What is ur super when u call super?
> > 
> > 
> > My understanding is that `super()` picks a certain evaluation order along which it tries to find an `__init__` to call, and then it call the first it finds. But if that `__init__` in turn calls `super().__init__(**kwargs)` then it will keep 'walking' along that evaluation order. So as long as every init in a mixin makes that call it should all work as intended. The mixins just need to be set up to not break this chain.
> 
> I agree with Johannes it would be better to add it in the Mixin `__init__` but if it is not possible because of inheritance problem we can put in in the DA **init**

Do not put it there. Mixins should be stateless

**Comment by JohannesMessner** [comment]
> > > > > > Do not add an `init` to the `Mixin`. DO that in the `init` of DocArray Base class
> > > > > 
> > > > > 
> > > > > Are we sure about this? It would assign some methods that don't exist there and break separation of concerns. It seems to be accepted practice to let mixins have inits and pass **kwargs between them: https://stackoverflow.com/a/50465583
> > > > 
> > > > 
> > > > multiple inheritance is not so easily supported in Python I think. What is ur super when u call super?
> > > 
> > > 
> > > My understanding is that `super()` picks a certain evaluation order along which it tries to find an `__init__` to call, and then it call the first it finds. But if that `__init__` in turn calls `super().__init__(**kwargs)` then it will keep 'walking' along that evaluation order. So as long as every init in a mixin makes that call it should all work as intended. The mixins just need to be set up to not break this chain.
> > 
> > 
> > I agree with Johannes it would be better to add it in the Mixin `__init__` but if it is not possible because of inheritance problem we can put in in the DA **init**
> 
> Do not put it there. Mixins should be stateless

But by moving it to the base class we don't avoid the state, we just move it to somewhere else. The only thing that changes is that the assignment happens somewhere where we don't know the thing that we assign...

But not willing to die in this hill, @anna-charlotte you can choose who to listen to ;)

**Comment by JoanFM** [comment]
> > > > > > > Do not add an `init` to the `Mixin`. DO that in the `init` of DocArray Base class
> > > > > > 
> > > > > > 
> > > > > > Are we sure about this? It would assign some methods that don't exist there and break separation of concerns. It seems to be accepted practice to let mixins have inits and pass **kwargs between them: https://stackoverflow.com/a/50465583
> > > > > 
> > > > > 
> > > > > multiple inheritance is not so easily supported in Python I think. What is ur super when u call super?
> > > > 
> > > > 
> > > > My understanding is that `super()` picks a certain evaluation order along which it tries to find an `__init__` to call, and then it call the first it finds. But if that `__init__` in turn calls `super().__init__(**kwargs)` then it will keep 'walking' along that evaluation order. So as long as every init in a mixin makes that call it should all work as intended. The mixins just need to be set up to not break this chain.
> > > 
> > > 
> > > I agree with Johannes it would be better to add it in the Mixin `__init__` but if it is not possible because of inheritance problem we can put in in the DA **init**
> > 
> > 
> > Do not put it there. Mixins should be stateless
> 
> But by moving it to the base class we don't avoid the state, we just move it to somewhere else. The only thing that changes is that the assignment happens somewhere where we don't know the thing that we assign...
> 
> But not willing to die in this hill, @anna-charlotte you can choose who to listen to ;)

What doesn't need state is the Mixin, not the class.

**Review by JoanFM — APPROVED** [review]

**Comment by JohannesMessner** [comment]
> > > > > > > > Do not add an `init` to the `Mixin`. DO that in the `init` of DocArray Base class
> > > > > > > 
> > > > > > > 
> > > > > > > Are we sure about this? It would assign some methods that don't exist there and break separation of concerns. It seems to be accepted practice to let mixins have inits and pass **kwargs between them: https://stackoverflow.com/a/50465583
> > > > > > 
> > > > > > 
> > > > > > multiple inheritance is not so easily supported in Python I think. What is ur super when u call super?
> > > > > 
> > > > > 
> > > > > My understanding is that `super()` picks a certain evaluation order along which it tries to find an `__init__` to call, and then it call the first it finds. But if that `__init__` in turn calls `super().__init__(**kwargs)` then it will keep 'walking' along that evaluation order. So as long as every init in a mixin makes that call it should all work as intended. The mixins just need to be set up to not break this chain.
> > > > 
> > > > 
> > > > I agree with Johannes it would be better to add it in the Mixin `__init__` but if it is not possible because of inheritance problem we can put in in the DA **init**
> > > 
> > > 
> > > Do not put it there. Mixins should be stateless
> > 
> > 
> > But by moving it to the base class we don't avoid the state, we just move it to somewhere else. The only thing that changes is that the assignment happens somewhere where we don't know the thing that we assign...
> > But not willing to die in this hill, @anna-charlotte you can choose who to listen to ;)
> 
> What doesn't need state is the Mixin, not the class.

I don't agree, the only 'state' we have is which implementation of `form_csv` we pick, and this is tied to the mixin, not anything else.

**Review by JoanFM — CHANGES_REQUESTED** [review]
sorry, not approved, let's raise Exception

**Review by samsja — CHANGES_REQUESTED** [review]
Looks good so far ! Next step is to make it a bit more generic so that it work for the all of the `from_xxx` classmethod

**Review by JoanFM — APPROVED** [review]

**Pull request merged**